ergm.py

```diff
@@ -187,6 +201,127 @@ def sample_network(self, sampling_method="NaiveMetropolisHastings", seed_network
+                 initial_thetas=None,
+                 initial_normalization_factor=None,
+                 seed_MCMC_proba=0.25):
+        super().__init__(n_nodes,
+                         network_statistics,
+                         is_directed,
+                         initial_thetas,
+                         initial_normalization_factor,
+                         seed_MCMC_proba)
+        self._all_weights = self._calc_all_weights()
+        self._normalization_factor = self._all_weights.sum()
+
+    # TODO: think of an alternative name? confusing with the static method
+    def _get_adj_mat_from_int(self, int_code: int):
+        return BruteForceERGM.construct_adj_mat_from_int(int_code, self._n_nodes, self._is_directed)
+
+    def _calc_all_weights(self):
+        if self._n_nodes > BruteForceERGM.MAX_NODES_BRUTE_FORCE:
+            raise ValueError(
+                f"The number of nodes {self._n_nodes} is larger than the maximum allowed for brute force "
+                f"calculations {BruteForceERGM.MAX_NODES_BRUTE_FORCE}")
+        num_pos_connects = self._n_nodes * (self._n_nodes - 1)
+        if not self._is_directed:
+            num_pos_connects //= 2
+        space_size = 2 ** num_pos_connects
+        all_weights = np.zeros(space_size)
+        for i in range(space_size):
+            cur_adj_mat = self._get_adj_mat_from_int(i)
+            all_weights[i] = super().calculate_weight(cur_adj_mat)
+        return all_weights
+
+    def calculate_weight(self, W: np.ndarray):
+        adj_mat_idx = BruteForceERGM.construct_int_from_adj_mat(W, self._is_directed)
+        return self._all_weights[adj_mat_idx]
+
+    def sample_network(self, sampling_method="Exact", seed_network=None, steps=0):
+        if sampling_method != "Exact":
+            raise ValueError("BruteForceERGM supports only exact sampling (this is its whole purpose)")
+
+        all_nets_probs = self._all_weights / self._normalization_factor
+        sampled_idx = np.random.choice(all_nets_probs.size, p=all_nets_probs)
+        return self._get_adj_mat_from_int(sampled_idx)
+
+    # TODO: these inherited ghost parameters are ugly (also in the sample_network method)
+    def fit(self, observed_network, n_networks_for_norm=100, n_mcmc_steps=500, verbose=True, optimization_options={}):
+        def nll(thetas):
+            model = BruteForceERGM(self._n_nodes, self._network_statistics, initial_thetas=thetas,
+                                   is_directed=self._is_directed)
+            return np.log(model._normalization_factor) - np.log(model.calculate_weight(observed_network))
+
+        def nll_grad(thetas):
+            model = BruteForceERGM(self._n_nodes, self._network_statistics, initial_thetas=thetas,
+                                   is_directed=self._is_directed)
+            observed_features = model._network_statistics.calculate_statistics(observed_network)
+            all_probs = model._all_weights / model._normalization_factor
```

Comment:
Why divide by `model._normalization_factor`?
Why not just do
` num_nets = len(model._all_weights)` or .size or something?